### PR TITLE
Make SDK `pyodide`/browser compatible out-of-the-box

### PR DESCRIPTION
**Pull request by haakonvt**
The first file any import into `cognite.client` hits is `cognite.client.__init__`. Thus, this is where we add our patching function to make sure we can, before the user actually imports anything, monkey-patch all the corners of the SDK to make it `pyodide` compatible (i.e. running in a browser in WebAssembly).

This allows us also to make authenticating the `CogniteClient` ridiculously easy if-and-only-if we pick up that the user is running in Fusion (where we pass all necessary auth details for the user to not worry about anything):
```py
from cognite.client import CogniteClient
client = CogniteClient()
```
...the magic trick is executed as part of patching by:
```py
if os.getenv("COGNITE_FUSION_NOTEBOOK") is not None:
    cc.config.global_config.default_client_config = FusionNotebookConfig()
```

### CODE REVIEW

File: cognite/client/utils/_pyodide_helpers.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+    # - DatapointsAPI.retrieve/retrieve_arrays/retrieve_dataframe:
+    cc._api.datapoints.PriorityThreadPoolExecutor = SerialPriorityThreadPoolExecutor  # type: ignore [assignment, misc]
+    cc._api.datapoints.as_completed = serial_as_completed  # type: ignore [assignment]
+
+
+def pyodide_http_adapter_send(self: Any, request: PreparedRequest, **kwargs: Any) -> Response:
+    # 'self' is actually `pyodide_http._requests.PyodideHTTPAdapter`, but we don't want it as a
+    # dependency just for the sake of mypy; after all, we are patching like monkeys 🙈
+    (response := self._old_send(request, **kwargs)).raw.version = ""
```
Comment (andeplane) [inline cognite/client/utils/_pyodide_helpers.py:87]: I think this can be simplified by changing how we test for raw.version instead of patching this.

Comment (haakonvt) [reply cognite/client/utils/_pyodide_helpers.py:87]: I find one usage of it in `cognite/client/_api_client.py`:
`http_protocol_version = ".".join(list(str(res.raw.version)))`

It is used to print out a debug message, i.e. `HTTP/{http_protocol_version}` -> `HTTP/1.1`. Let's say I change it to a try-except, which protocol / what message should be logged instead?

Comment (haakonvt) [reply cognite/client/utils/_pyodide_helpers.py:87]: I'll push a change, then feel free to correct the wording correctness 😉

File: cognite/client/utils/_pyodide_helpers.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    def __enter__(self) -> SerialPriorityThreadPoolExecutor:
+        return self
+
+    def __exit__(
+        self,
+        type: Optional[type[BaseException]],
+        value: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> None:
+        self.shutdown()
```
Comment (erlendvollset) [inline cognite/client/utils/_pyodide_helpers.py:141]: I would suggest moving these into the concurrency module along with the PriorityThreadPoolExecutor, and toggling their usage with the same ConcurrencySettings option. Makes it easier to follow if it's done the same way for the different implementations.

File: cognite/client/utils/_pyodide_helpers.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        self._OAuthInteractive__redirect_port = redirect_port
+        self._OAuthInteractive__app = PublicClientApplication(client_id=client_id, authority=authority_url)
+
+    @staticmethod
+    def _load_payload_from_jwt(token: str) -> Dict[str, Any]:
+        """A JWT has the structure: `<header>.<payload>.<signature>`. This method returns payload as a dict"""
+        _, payload, _ = token.split(".")
+        if remainder := len(payload_bytes := payload.encode()) % 4:
+            payload_bytes += b"=" * (4 - remainder)
+        return json.loads(urlsafe_b64decode(payload_bytes))
```
Comment (erlendvollset) [inline cognite/client/utils/_pyodide_helpers.py:188]: IMO if you're using the variable outside the scope of the if-statement, it's easier to follow if you define it upfront.
```suggestion
        payload_bytes = payload.encode()
        if remainder := len(payload_bytes) % 4:
            payload_bytes += b"=" * (4 - remainder)
        return json.loads(urlsafe_b64decode(payload_bytes))
```

File: cognite/client/utils/_pyodide_helpers.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        # Note: Death to name mangling:
+        self._OAuthCredentialProviderWithTokenRefresh__access_token = token
+        self._OAuthCredentialProviderWithTokenRefresh__token_refresh_lock = contextlib.nullcontext()
+        self._OAuthCredentialProviderWithTokenRefresh__access_token_expires_at = payload["exp"]
+
+        self._OAuthInteractive__authority_url = authority_url
+        self._OAuthInteractive__client_id = client_id
+        self._OAuthInteractive__scopes = [f"{base_url}/.default"]
+        self._OAuthInteractive__redirect_port = redirect_port
+        self._OAuthInteractive__app = PublicClientApplication(client_id=client_id, authority=authority_url)
```
Comment (erlendvollset) [inline cognite/client/utils/_pyodide_helpers.py:180]: Not a big fan of this. 

Someone who goes and changes the implementation of OAuthCredentialProviderWithTokenRefresh or OAuthInteractive won't necessarily be aware of this class accessing their private attributes and making assumptions on their implementations. In this case I would say it's better to just copy the whole implementation into this module. 

Especially considering that we don't have tests covering this which would catch this thing breaking.

File: cognite/client/utils/_pyodide_helpers.py (hunk 1 above)
Comment (andeplane) [reply cognite/client/utils/_pyodide_helpers.py]: I think that was the one that crashed (hence the fix), yep

File: cognite/client/utils/_pyodide_helpers.py (hunk 4 above)
Comment (haakonvt) [reply cognite/client/utils/_pyodide_helpers.py]: Removed altogether while we wait on a solution for proper token refresh handling in coop with Fusion

File: cognite/client/utils/_pyodide_helpers.py (hunk 3 above)
Comment (haakonvt) [reply cognite/client/utils/_pyodide_helpers.py]: Entire class removed, see other comment

File: cognite/client/utils/_pyodide_helpers.py (hunk 2 above)
Comment (haakonvt) [reply cognite/client/utils/_pyodide_helpers.py]: Moved what could be moved without too much trouble. Patching now looks unified:

```py
# - Set all usage of thread pool executors to use dummy/serial-implementations:
cc.utils._concurrency.ConcurrencySettings.executor_type = "mainthread"
cc.utils._concurrency.ConcurrencySettings.priority_executor_type = "mainthread"
```

Skipped: Unable to use the same singleton pattern due to various logic calling on the pool to shut down explicitly. It seems like a significant risk to change that logic atm.

Comment (haakonvt) [reply cognite/client/utils/_pyodide_helpers.py]: Also moved the usage of `concurrent.future.as_completed` to a pool method, that way we can forget about patching it 😄

**Comment by haakonvt** [comment]
@erlendvollset is this one good to go? 😄

File: CHANGELOG.md [hunk 5]
```diff
@@ -17,6 +17,10 @@ Changes are grouped as follows
 - `Fixed` for any bug fixes.
 - `Security` in case of vulnerabilities.
 
+## [5.5.3] - 16-02-23
```
Comment (erlendvollset) [inline CHANGELOG.md:20]: Adding new functionality merits a minor bump
```suggestion
## [5.6.0] - 16-02-23
```

**Review by erlendvollset — APPROVED** [review]
Approved! But I think a minor bump is more appropriate.

**Pull request merged**